### PR TITLE
feat: run Test tasks in the runner

### PR DESCRIPTION
**Pull request by ckerr**
FIxes https://github.com/electron/bugbot/issues/104.

### CODE REVIEW

File: modules/runner/src/runner.ts [hunk 1]
```diff
@@ -75,63 +141,74 @@ export class Runner {
...
   public start(): void {
+    const d = debug(`${DebugPrefix}:start`);
+
     this.stop();
     d('runner:start', `interval is ${this.pollIntervalMs}`);
-    this.interval = setInterval(() => this.pollSafely(), this.pollIntervalMs);
+    this.pollInterval = setInterval(
+      () => this.pollSafely(),
+      this.pollIntervalMs,
+    );
```
Comment (erickzhao) [inline modules/runner/src/runner.ts:151]: question: is there a difference between adding a wrapper function and just doing this?
```suggestion
    this.pollInterval = setInterval(this.pollSafely,
      this.pollIntervalMs,
    );
```

File: modules/runner/src/runner.ts [hunk 2]
```diff
@@ -75,63 +141,74 @@ export class Runner {
...
 
+  private async pickNextJob(): Promise<JobId | undefined> {
+    const d = debug(`${DebugPrefix}:pickNextJob`);
+
+    const ids = await this.fetchAvailableJobIds();
+    if (!ids.length) return;
+
+    const idx = randomInt(0, ids.length);
+    const [id] = ids.splice(idx, 1);
+    d('jobs %o picked %s', ids, id);
```
Comment (erickzhao) [inline modules/runner/src/runner.ts:204]: nit: we should specify that the first `%o` is jobs remaining

File: modules/runner/src/runner.ts [hunk 3]
```diff
@@ -5,23 +5,94 @@ import which from 'which';
...
+
+  public addLogData(data: string) {
+    // save the URL to safeguard against this.jobId being cleared at end-of-job
+    const log_url = new URL(`api/jobs/${this.job.id}/log`, this.brokerUrl);
+    this.logBuffer.push(data);
+    if (!this.logTimer) {
+      this.logTimer = setTimeout(
+        () => void this.sendLogDataBuffer(log_url),
+        this.logIntervalMs,
+      );
```
Comment (erickzhao) [inline modules/runner/src/runner.ts:69]: question: is `logIntervalMs` long enough that we won't run into any race conditions with this asynchronous code?

File: modules/shared/src/interfaces.ts [hunk 4]
```diff
@@ -23,32 +23,39 @@ export type RunnerId = string;
...
 
-export type BisectRange = [Version, Version];
-const BisectRangePredicate = ow.array.length(2).ofType(VersionPredicate);
+export type VersionRange = [Version, Version];
+const VersionRangePredicate = ow.array
+  .ofType(VersionPredicate)
+  .length(2)
+  .is((arr) => {
+    const [v1, v2] = arr;
+    if (semver.gte(v1, v2)) return `expected ${v1} to be less than ${v2}`;
```
Comment (erickzhao) [inline modules/shared/src/interfaces.ts:37]: question: does `semver.gte` the `nightly` vs `beta` edge-case?

File: integration-tests/test/broker-runner.test.ts [hunk 5]
```diff
@@ -47,7 +49,7 @@ describe('runner', () => {
   }
 
   afterEach(async () => {
-    brokerServer.stop();
+    await brokerServer.stop();
     await runner.stop();
```
[the remark this answers is not in the document]
Comment (ckerr) [reply integration-tests/test/broker-runner.test.ts:53]: I might be confused, but I think [it does](https://github.com/electron/bugbot/blob/31f2b6494685e934b75fb77e031d660f08bd1fbf/modules/broker/src/server.ts#L336)?

[the remark this answers is not in the document]
Comment (erickzhao) [reply integration-tests/test/broker-runner.test.ts:53]: The broker does, but the runner doesn't: https://github.com/electron/bugbot/blob/31f2b6494685e934b75fb77e031d660f08bd1fbf/modules/runner/src/runner.ts#L84

File: modules/runner/src/runner.ts (hunk 2 above)
Comment (ckerr) [reply modules/runner/src/runner.ts]: clarified in f2bc49481861b55e96b28894fc07f1481617b8c4

File: modules/runner/src/runner.ts [hunk 6]
```diff
@@ -5,23 +5,94 @@ import which from 'which';
...
 import { env, envInt } from '@electron/bugbot-shared/build/env-vars';
 
 import { parseFiddleBisectOutput } from './fiddle-bisect-parser';
 
-const d = debug('runner');
+const DebugPrefix = 'runner' as const;
+
+class Task {
+  private logTimer: ReturnType<typeof setTimeout>;
+  public readonly logBuffer: string[] = [];
```
[the remark this answers is not in the document]
Comment (ckerr) [reply modules/runner/src/runner.ts]: Yes. I had it public during the extract-Task refactor because some parts of Runner still touched it, and then I forgot to make it private when done.

Fixed in 8d3eff760e3d76726234bfd3e13e3eb79bc6940b

File: integration-tests/test/broker-runner.test.ts [hunk 7]
```diff
@@ -160,9 +164,93 @@ describe('runner', () => {
...
     it('includes the commit range to job.log', () => {
       const log = task.getRawLog();
-      const [a, b] = bisect_range;
+      const [a, b] = version_range;
       const url = `https://github.com/electron/electron/compare/v${a}...v${b}`;
       expect(log).toMatch(url);
     });
   });
+
+  function createTestTask(job: Partial<TestJob> = {}) {
```
[the remark this answers is not in the document]
Comment (ckerr) [reply integration-tests/test/broker-runner.test.ts]: Sounds good. Moved in 22c1db6c95e1125c5a86c0e5bc2c80c87050bf1b

File: modules/shared/src/interfaces.ts (hunk 4 above)
Comment (ckerr) [reply modules/shared/src/interfaces.ts:37]: Ugh, no. I'll clone that code over from Fiddle.

BTW https://github.com/electron/fiddle/pull/764/files#diff-682f619cb0bef8a3eabc1a9b2a2bccc77030e45ef7662525427a690009a2a2f5 updates that code to work for alpha releases as well :slightly_smiling_face:

File: modules/runner/src/runner.ts (hunk 3 above)
Comment (ckerr) [reply modules/runner/src/runner.ts:69]: logTimer and logBuffer are both updated in the same block of code in `sendLogDataBuffer()` _before_ the function yields for the `await fetch` a few lines below, and `addLogData()` is a synchronous function, so I don't think I see any race condition here regardless of the interval?

Maybe there's something I've overlooked. Is there something you have in mind?

File: modules/runner/src/runner.ts (hunk 1 above)
Comment (ckerr) [reply modules/runner/src/runner.ts:151]: Yes, the latter generates an eslint warning for `this`:

File: integration-tests/test/broker-runner.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ckerr) [reply integration-tests/test/broker-runner.test.ts]: Yep, you are right. Fixed in 0e339ab82

File: modules/runner/src/runner.ts (hunk 1 above)
Comment (ckerr) [reply modules/runner/src/runner.ts]: Updated in eb87705f50ac13057e3c0d2b949fb3cf72c7e0a2

File: modules/shared/src/interfaces.ts (hunk 4 above)
Comment (ckerr) [reply modules/shared/src/interfaces.ts]: Fixed in 42544d62a2007c1c517438d3411e75eb76d0d579

File: modules/runner/src/runner.ts (hunk 3 above)
Comment (erickzhao) [reply modules/runner/src/runner.ts:69]: I didn't look too deeply into what happens in the `PUT` call, but if there's no problem with these happening out of order somehow (since call N+1 won't wait for call N), this seems fine to me.

I think you made that fairly explicit with the `void` operator, but just double-checking!

**Review by erickzhao — APPROVED** [review]
